### ISSUE TITLE
Replacing explicit ShortName with Label in Chunk types

### ISSUE THREAD

**Issue by elwazana** [JacquesCarette/Drasil#695]
### Chunk
- [ ] AssumpChunk
- [ ] Change
- [ ] Citation **!**
- [ ] Eq
- [ ] GenDefn
- [ ] Goal **!**
- [ ] InstanceModel
- [ ] PhysSystDesc **!**
- [ ] Relation **!**
- [ ] ReqChunk
- [ ] Theory

In PR #693 @JacquesCarette  and I discussed replacing the explicit shortname in some Chunk types with a `Label`, I've created this issue just to keep track of which ones need to be replaced.

Also as a note the (**!**) next to some are due to them having an instance of `HasShortName` but posses no `ShortName`, i.e:
![image](https://user-images.githubusercontent.com/35268928/41724817-65c3f072-753c-11e8-8c66-d08ca97645cf.png)
@JacquesCarette Should I add a `Label` to these as well?

**Comment by JacquesCarette** [comment]
Goal and PhysSystDesc should definitely have labels.

Relation is a horrible hack. But that is a separate thing. The way it is abused right now means that it should likely have a label as well.

If a Citation is a bibliography entry, then yes, that too should have a label.

**Comment by elwazana** [comment]
@JacquesCarette Hello Dr. Carette, I missed Concept/Core.hs when I was creating the task list above. Does that need a `Label` as well? It currently has a `Maybe Shortname`.

**Comment by Mornix** [comment]
`ConceptChunk` gained the `Maybe ShortName` last week as part of the work to de-embed ALUR (and make `ConceptChunk` referable. 
It is currently unused. All constructors at the moment simply set it as `Nothing`.

**Comment by JacquesCarette** [comment]
Right. That should be upgraded to a `Maybe Label`. We can refine later, but at least that should allow both @elwazana and @Mornix to proceed.

**Comment by elwazana** [comment]
@JacquesCarette Hello Dr. Carette, I've tried incorporating `Maybe Label` but whenever I get to the instance of `HasLabel` I can't figure out how to use a `Lens'` to unwrap the `Maybe` from the `Maybe Label`. 

I've looked through the other types to see if a Lens was used over a `Maybe` type and the only way around this I could find is to create a:
``` haskell
class HasMaybeLabel c where
  getMaybeLabel :: Lens' c (Maybe Label)
```

But this solution seems bit smelly to me, is there another solution? If not can I get a go ahead on this solution?

**Comment by JacquesCarette** [comment]
No, that's right, you have to create a Lens to a Maybe Label.

**Comment by elwazana** [comment]
@JacquesCarette I implemented the above to correct the `Maybe Label` error in the instance of `HasLabel`, after that, I tried to fix the `HasShortName` instance for `ConceptChunk` so that it retrieves the `shortname` from the `Label`. When I did that I got this error:
![image](https://user-images.githubusercontent.com/35268928/41745359-b014968a-7574-11e8-9419-50999c47116c.png)

So to fix that I added an instance of `HasShortName` specifically for `Maybe Label`:
![image](https://user-images.githubusercontent.com/35268928/41745332-9810f060-7574-11e8-9041-b9e112749622.png)

But this resulted in this error:
![image](https://user-images.githubusercontent.com/35268928/41745313-894c6c76-7574-11e8-95ef-3ee71a1fb131.png)

I've discussed possible solutions with @niazim3 but nothing we come up with seems to even change the error.

**Comment by JacquesCarette** [comment]
So you won't be able to build a proper instance of `HasShortName` for `ConceptChunk` out of `getMaybeLabel`. This is because it might be `Nothing`, which is not a valid shortname!

In other words, the handling of label/shortname for `ConceptChunk` is going to have to be coded differently.  Right now, since it's not used, simply don't worry about it, we can let @Mornix deal with it later.

**Comment by niazim3** [comment]
Is there an example of a time where a `Label` would not be desirable when building a `ConceptChunk`? If not, can Concept/Core.hs take a `Label` instead of a `Maybe Label`?

**Comment by Mornix** [comment]
From when I was looking at the current`ConceptChunk` uses, it looks like the type backs a lot of things in Data.Drasil and I believe more localized concepts in each individual example. These currently aren’t referable. (I’m not sure if they should be referable in the future /cc @JacquesCarette)

**Comment by JacquesCarette** [comment]
Right.  `ConceptChunk` is very basic in our hierarchy, and often corresponds to purely internal knowledge that does not necessarily even appear in documents. Those do not need to be referable.

Eventually I think we will end up splitting our uses of `ConceptChunk` into more meaningful sub-categories, and then some of the sub-categories will have labels, and the others will not.  But we're not quite ready for that yet.

**Comment by niazim3** [comment]
Would it be a good idea to work on the other data types as discussed in previous issues and meetings in the meanwhile (ignoring the `Maybe` problem brought up by Concept/Core)?

Perhaps another issue can be made for that as this is commented out (to ensure the implementations for the other data types are correct)?

Or is this issue being in the midst of being worked on/close to being finished that we can wait on completion of this whole issue?

**Comment by JacquesCarette** [comment]
Yes, getting many of the data-types which need this work done now is a good idea.

`ConceptChunk` is not close enough to rely on. So it would be best if this could proceed independently.

**Comment by elwazana** [comment]
@JacquesCarette Also with regards to `Section`:
![image](https://user-images.githubusercontent.com/35268928/41938165-4be1cd5a-7960-11e8-8e1c-3f5d21d0e591.png)
Would it make more sense to have `Section` as a record type? Or should it be kept as is for the design?

**Comment by JacquesCarette** [comment]
We discussed removing the instance for `Concept` yesterday. Not `Contents`.

Although grabbing a label from a `Contents` might eventually make more sense than directly getting the shortname.

Yes, it makes sense to make `Section` a record type, as that can make accessing parts easier.

**Comment by elwazana** [comment]
@JacquesCarette I was referring to issue #629 task 4 of @niazim3 checklist when I mentioned removing the instances of `Contents`, unless that has been changed?

**Comment by JacquesCarette** [comment]
Ah, that makes sense.  Right, `Content` will be 'raw' and `LabelledContent` will have the `Label` and `Content` will not.  I had forgotten - perfect.

**Comment by elwazana** [comment]
@JacquesCarette Hello Dr. Carette, I've removed the instances of  `Contents` and I had a question about what should be done with the referencing functions in `References.hs`:
![image](https://user-images.githubusercontent.com/35268928/42228971-4ac49724-7eb3-11e8-98e3-0c2ccbd17730.png)

There are two issues here:

1. `acroTest` and `find'` use `makeRef` to work, however, `makeRef` requires a type with a `HasShortName` and `Referable` instance which `Contents` doesn't have anymore.

2. With regards to `find'` should it be changed to:
``` Haskell
find' :: LabelledContent -> [LabelledContent] -> LabelledContent
find' = ~~~~~
```

**Comment by niazim3** [comment]
1. Re: Since `LabelledContent` is now an instance of `HasShortName` and `Referable`, I think it can replace `Contents`.

**Comment by JacquesCarette** [comment]
I agree with @niazim3 , it does seem like `LabelledContent` should fit the bill.

**Comment by elwazana** [comment]
@JacquesCarette Hello Dr. Carette I was able to integrate `LabelledContent` in place of `Contents` for the referencing functions and that runs fine now. But I've run into this problem that I'm unsure how to solve.

![image](https://user-images.githubusercontent.com/35268928/42279746-d04ee142-7f6c-11e8-89f4-c4c2bff5a1d5.png)

The problem is that `convertRel` uses `ec` which is a constructor for `Eq`, this was adjusted to take in a `Label` instead of a `ShortName`, but fixing that here to take a `Label` causes the following to break:

![image](https://user-images.githubusercontent.com/35268928/42279802-f9392c84-7f6c-11e8-8c4c-5565ccd5b568.png)

I was able to fix the issue by doing this:
![image](https://user-images.githubusercontent.com/35268928/42280310-5fcd8cfa-7f6e-11e8-92ff-a822845293e6.png)

However, this seems a bit off to me. I'm not sure if this is becuase `relToQD` is a hack (issue #628), or its becuase of my fix. So I just wanted to get some feedback on it before getting too far ahead.

**Comment by JacquesCarette** [comment]
Both `relToQD` and `convertRel` are awful hacks. Your change is 'optimal' given the constraints.

The correct fix involves ripping out `ExprRelat` altogether and creating a proper class for 'definitions'. Right now, our definitions are done implicitly through the relational subset of `Expr`, and is super-fragile.  However, it will be needed once we move to doing code generation for ODEs, for example.

It would be nice if we could get to at least fixing the `relToQD` hack this summer.

**Comment by niazim3** [comment]
As of the most recent commit, build is breaking at the following error:
```
.\code\Data\Drasil\Utils.hs:139:18: error:
    • No instance for (HasShortName Contents)
        arising from a use of ‘makeRef’
    • In the first argument of ‘(.)’, namely ‘makeRef’
      In the expression: (makeRef . Definition . f)
      In an equation for ‘refFromType’:
          refFromType f = (makeRef . Definition . f)
```
. It originally looked like all that needs to be done is convert the `Contents` created with the call on `Definition` to a `LabelledContent` (may still be the case). 

This `refFromType` is used in the examples on `[QDefinitions]` and `[RelationConcept]` (e.g. 
https://github.com/JacquesCarette/Drasil/blob/b3938fb5e7013e5165a0be8c1d0641bf08a7f10f/code/Example/Drasil/GlassBR/Body.hs#L617-L623) to tack on "DD" or "T" (depending on if the element `Definition` is a `Theory` or `Data`):
https://github.com/JacquesCarette/Drasil/blob/b3938fb5e7013e5165a0be8c1d0641bf08a7f10f/code/Language/Drasil/Reference.hs#L220
https://github.com/JacquesCarette/Drasil/blob/b3938fb5e7013e5165a0be8c1d0641bf08a7f10f/code/Language/Drasil/Reference.hs#L230.,since all `makeRef` is expecting is anything that is an instance of `HasShortName` and `Referable`, which anything with a `Label` will be.

If `RelationConcept` becomes an instance of `HasLabel`, there will be no need for any conversion to a `Contents` or `LabelledContent` or for appending "`T:`" or "`DD:`" to any labels.

**Comment by niazim3** [comment]
Having said that, #628 mentions that `RelationConcept` is too raw/unstable to be augmented; therefore, TMs, IMs, and GDs should build off of `RelationConcept` (which `InstanceModel` and `GenDefn` already does:
https://github.com/JacquesCarette/Drasil/blob/b3938fb5e7013e5165a0be8c1d0641bf08a7f10f/code/Language/Drasil/Chunk/InstanceModel.hs#L32-L44
https://github.com/JacquesCarette/Drasil/blob/b3938fb5e7013e5165a0be8c1d0641bf08a7f10f/code/Language/Drasil/Chunk/GenDefn.hs#L17-L24).
Is there a reason `TheoryChunk` doesn't https://github.com/JacquesCarette/Drasil/blob/b3938fb5e7013e5165a0be8c1d0641bf08a7f10f/code/Language/Drasil/Chunk/Theory.hs#L32-L42?

**Comment by JacquesCarette** [comment]
`TheoryChunk` is good at it is. It is a collection of lots of things.
`GenDefn` and `InstanceModel` are pretty good, but still a bit hacky. Mostly because `RelationConcept` is somewhat broken. It should really be a generalization of `QDefinition`. This hack is similar to the one afflicting `relToQD`.

**Comment by JacquesCarette** [comment]
Regarding the previous comment: even if `RelationConcept` is sub-optimal (but just in the 'fact' that it has a relation instead of a defining equation, which it usually has, but encoded as a relation), it is still the right 'base' on which to build lots of stuff.  And thus adding a `Label` to it is a good idea.

**Comment by niazim3** [comment]
If a `Label` is added to `RelationConcept`, would that `Label` be propagated to the `InstanceModel` and `GenDefn`s that build off of the `RelationConcept`?
Or would those data instances have their own `Label`?

**Comment by JacquesCarette** [comment]
I think that they would be separate labels. Hmm, I think I was assuming `Maybe Label` for `RelationConcept`, not just `Label`.

**Comment by JacquesCarette** [comment]
Just a note for @Nathaniel-Hu and @muhammadaliog3 that your work of finding all the data, classes and instances will have an impact on this issue. The underlying reason it's still not fixed is that our concept matrix isn't quite "right".
